### PR TITLE
Use `resolveRequest` to compare symlinks, rather than `fs.realpathSync` (fix pnpm on Vercel)

### LINKED ISSUE #20948: ENOENT build error with Vercel + pnpm
**What version of Next.js are you using?**

10.0.5

**What version of Node.js are you using?**

14.15.3

**What browser are you using?**

Edge (Chromium)

**What operating system are you using?**

macOS

**How are you deploying your application?**

Vercel

**Describe the Bug**

Context:
* https://github.com/vercel/next.js/issues/16471#issuecomment-736095201
* https://github.com/vercel/vercel/discussions/3896#discussioncomment-131265
* https://github.com/elliottsj/nextjs-vercel-pnpm-issue

As discussed above and described by @vasiliicuhar [here](https://github.com/vercel/next.js/issues/16471#issuecomment-757478987), the latest Next.js does not build on Vercel when using pnpm, due to:

```
Error: ENOENT: no such file or directory, lstat '/vercel/workpath0/app/url'
--
15:28:10.448 | at realpathSync (fs.js:1643:7)
15:28:10.448 | at handleExternals (/vercel/workpath0/app/node_modules/.pnpm/next@10.0.5_react-dom@17.0.1+react@17.0.1/node_modules/next/dist/build/webpack-config.js:66:21)
```

**Expected Behavior**

The error above should not occur, and the project should build successfully.

**To Reproduce**

A full reproduction is described here: https://github.com/elliottsj/nextjs-vercel-pnpm-issue

### PR DESCRIPTION
**Pull request by elliottsj**
Ported from https://github.com/vercel/next.js/pull/16369

This fixes #20948 by avoiding the error:

```
Error: ENOENT: no such file or directory, lstat '/vercel/workpath0/app/url'
--
15:28:10.448 | at realpathSync (fs.js:1643:7)
15:28:10.448 | at handleExternals (/vercel/workpath0/app/node_modules/.pnpm/next@10.0.5_react-dom@17.0.1+react@17.0.1/node_modules/next/dist/build/webpack-config.js:66:21)
```

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | elliottsj/next.js resolve-request | Change |
| - | - | - | - |
| buildDuration | 10.2s | 10.4s | ⚠️ +111ms |
| nodeModulesSize | 80.9 MB | 80.9 MB | ⚠️ +7.96 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | elliottsj/next.js resolve-request | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 1.975 | 1.969 | -0.01 |
| / avg req/sec | 1265.55 | 1269.88 | +4.33 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.226 | 1.271 | ⚠️ +0.04 |
| /error-in-render avg req/sec | 2038.85 | 1966.49 | ⚠️ -72.36 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | elliottsj/next.js resolve-request | Change |
| - | - | - | - |
| 677f882d2ed8..89aa.js gzip | 13.1 kB | 13.1 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-dba4ec1..27c0.js gzip | 6.63 kB | 6.63 kB | ✓ |
| webpack-50be..df5b.js gzip | 751 B | 751 B | ✓ |
| Overall change | 59.4 kB | 59.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | elliottsj/next.js resolve-request | Change |
| - | - | - | - |
| polyfills-81..14d7.js gzip | 31.2 kB | 31.2 kB | ✓ |
| Overall change | 31.2 kB | 31.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | elliottsj/next.js resolve-request | Change |
| - | - | - | - |
| _app-b6fc6bc..222c.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-e2ffa..0f3f.js gzip | 3.46 kB | 3.46 kB | ✓ |
| hooks-010c20..8411.js gzip | 887 B | 887 B | ✓ |
| index-bbee2f..528b.js gzip | 227 B | 227 B | ✓ |
| link-d979103..c5ff.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..bf84.js gzip | 303 B | 303 B | ✓ |
| withRouter-a..5826.js gzip | 302 B | 302 B | ✓ |
| Overall change | 8.09 kB | 8.09 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | elliottsj/next.js resolve-request | Change |
| - | - | - | - |
| _buildManifest.js gzip | 323 B | 323 B | ✓ |
| Overall change | 323 B | 323 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | elliottsj/next.js resolve-request | Change |
| - | - | - | - |
| index.html gzip | 614 B | 614 B | ✓ |
| link.html gzip | 619 B | 619 B | ✓ |
| withRouter.html gzip | 606 B | 606 B | ✓ |
| Overall change | 1.84 kB | 1.84 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | elliottsj/next.js resolve-request | Change |
| - | - | - | - |
| buildDuration | 12.1s | 12.1s | ⚠️ +35ms |
| nodeModulesSize | 80.9 MB | 80.9 MB | ⚠️ +7.96 kB |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | elliottsj/next.js resolve-request | Change |
| - | - | - | - |
| 677f882d2ed8..89aa.js gzip | 13.1 kB | 13.1 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-dba4ec1..27c0.js gzip | 6.63 kB | 6.63 kB | ✓ |
| webpack-50be..df5b.js gzip | 751 B | 751 B | ✓ |
| Overall change | 59.4 kB | 59.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | elliottsj/next.js resolve-request | Change |
| - | - | - | - |
| polyfills-81..14d7.js gzip | 31.2 kB | 31.2 kB | ✓ |
| Overall change | 31.2 kB | 31.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | elliottsj/next.js resolve-request | Change |
| - | - | - | - |
| _app-b6fc6bc..222c.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-e2ffa..0f3f.js gzip | 3.46 kB | 3.46 kB | ✓ |
| hooks-010c20..8411.js gzip | 887 B | 887 B | ✓ |
| index-bbee2f..528b.js gzip | 227 B | 227 B | ✓ |
| link-d979103..c5ff.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..bf84.js gzip | 303 B | 303 B | ✓ |
| withRouter-a..5826.js gzip | 302 B | 302 B | ✓ |
| Overall change | 8.09 kB | 8.09 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | elliottsj/next.js resolve-request | Change |
| - | - | - | - |
| _buildManifest.js gzip | 323 B | 323 B | ✓ |
| Overall change | 323 B | 323 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | vercel/next.js canary  | elliottsj/next.js resolve-request | Change |
| - | - | - | - |
| _error.js | 1 MB | 1 MB | ✓ |
| 404.html | 2.67 kB | 2.67 kB | ✓ |
| hooks.html | 1.92 kB | 1.92 kB | ✓ |
| index.js | 1 MB | 1 MB | ✓ |
| link.js | 1.06 MB | 1.06 MB | ✓ |
| routerDirect.js | 1.05 MB | 1.05 MB | ✓ |
| withRouter.js | 1.05 MB | 1.05 MB | ✓ |
| Overall change | 5.18 MB | 5.18 MB | ✓ |

</details>

</details>
Commit: 06b8e96372b840c9a421a60aa24a4e8e7b8ec012

File: packages/next/compiled/resolve/index.js [hunk 1]
```diff
@@ -1 +1 @@
...
...ilter){i=t.packageFilter(i,e)}return{pkg:i,dir:e}}function loadAsDirectorySync(e){var r=a.join(v(h,e,t),"/package.json");if(s(r)){try{var n=l(r,"UTF8");var i=JSON.parse(n)}catch(e){}if(i&&t.packageFilter){i=t.packageFilter(i,e)}if(i&&i.main){if(typeof i.main!=="string"){var o=new TypeError("package “"+i.name+"” `main` must be a string");o.code="INVALID_PACKAGE_MAIN";throw o}if(i.main==="."||i.main==="./"){i.main="index"}try{var u=loadAsFileSync(a.resolve(e,i.main));if(u)return u;var c=loadAsDirectorySync(a.resolve(e,i.main));if(c)return c}catch(e){}}}return loadAsFileSync(a.join(e,"/index"))}function loadNodeModulesSync(e,r){var n=function(){return d(e,r,t)};var i=y?y(e,r,n,t):n();for(var o=0;o<i.length;o++){var s=i[o];if(_(a.dirname(s))){var u=loadAsFileSync(s);if(u)return u;var l=loadAsDirectorySync(s);if(l)return l}}}}},747:e=>{"use strict";e.exports=require("fs")},622:e=>{"use strict";e.exports=require("path")}};var r={};function __webpack_require__(t){if(r[t]){return r[t].exports}var n=r[t]={exports:{}};var i=true;try{e[t](n,n.exports,__webpack_require__);i=false}finally{if(i)delete r[t]}return n.exports}__webpack_require__.ab=__dirname+"/";return __webpack_require__(305)})();
```
Comment (jrmcj) [inline packages/next/compiled/resolve/index.js:1]: 359778

**Review by jrmcj — APPROVED** [review]

**Comment by merceyz** [comment]
In https://github.com/vercel/next.js/pull/19518 this function is removed and replaced with `require.resolve` which probably makes this unnecessary

**Comment by elliottsj** [comment]
@merceyz Awesome. Looks like your PR was merged; I'll test it out soon to see if the issue is resolved.

**Comment by elliottsj** [comment]
Appears to be [fixed](https://github.com/vercel/next.js/issues/20948#issuecomment-758171176) 🎉

**Pull request closed without merging**